designs/spot-consolidation.md

```diff
@@ -0,0 +1,105 @@
+* 👎 Tuning this value to migitage both marginal-improvement-churn and lowest-price-regression is challenging, and desired values may conflict.
+
+Note: Regardless of the decision made to solve the spot consolidation problem, we’d likely want to implement a price improvement in the future to prevent consolidation from interrupting nodes to make marginal improvements.
+
+### 3. Optimistically Launch Replacements
+
+Karpenter could optimistically launch new EC2 instances to probe PCO decision making. If PCO launches a cheaper instance type from a CreateFleet request that include the current instance type, we know that the instance is both lower price and lower interruption rate. If the PCO algorithm selects the current instance type of the running instance, Karpenter would terminate the new instance and leave the current instance running. This approach would need to include an optimistic launch cache to minimize the number of optimistic launches.
+
+* 👍👍👍 Prevents consolidation from regressing to a lowest price allocation strategy
+* 👎 May not make sense for all cloud providers
+* 👎 May consolidate to make marginal improvements, resulting in unnecessary disruption
+* 👎 Optimistic launches may confuse customers looking deeply at their bill or audit logs
+* 👎 Bugs with this mechanism are higher risk, since real instances are being launched.
+
+### 4. Minimum Node Lifetime
+
+Karpenter could skip consolidation actions if the instance had been launched recently. This wouldn’t prevent the allocation strategy from regressing to lowest price, but given sufficient time (e.g. 4 hours), prevents the worst case scenario of repeated interruption.
+
+There are multiple ways to model this time based constraint. Minimum node lifetime would prevent any consolidation of a node under a given age. Alternatively, this could be modeled as minimum pod lifetime, where consolidation would not terminate a node until all pods are above a specified age. This is the most direct approach to mitigate the problem statement: if there’s too much interruption, reduce frequency. We should delay (if ever) enabling customers from being able to configure this. 
+
+Alternatively, we could rely on ttlSecondsAfterUnderutilized, where Karpenter would detect a potential consolidation and wait for the ttl before taking action. This is somewhat unnatural, since we’ve seen customers configure the similar ttlSecondsAfterEmpty in the range of 30-180 seconds, which isn’t enough time to avoid regression to lowest price. It’s likely that customers would want one value (e.g. 30s) to handle empty nodes and another value to prevent repeated interruption (e.g. 4h). Further, given that we’re building ttlSecondsAfterUnderutilized to meet other cases, customers would be required to reason about both minimum node lifetime and ttlSecondsAfterUnderutilized. 
```

Comment:
```suggestion
Alternatively, we could rely on ttlSecondsAfterUnderutilized, where Karpenter would detect a potential consolidation and wait for the ttl before taking action. This is somewhat unnatural, since we’ve seen customers configure the similar ttlSecondsAfterEmpty in the range of 30-180 seconds, which isn’t enough time to avoid regression to lowest price. It’s likely that customers would want one value (e.g. 30s) to handle empty nodes and another value to prevent repeated interruption (e.g. 4h). Further, given that we’re building consolidateAfter to meet other cases, customers would be required to reason about both minimum node lifetime and consolidateAfter. 
```